File: app/src/main/java/im/dacer/kata/SettingsActivity.java
/*
 * The MIT License (MIT)
 * Copyright (c) 2016 baoyongzhang <baoyz94@gmail.com>
 */
package im.dacer.kata;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SwitchCompat;
import android.view.MenuItem;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.TextView;

import im.dacer.kata.R;
import im.dacer.kata.core.BigBang;
import im.dacer.kata.core.action.CopyAction;
import im.dacer.kata.service.ListenClipboardService;
import com.baoyz.treasure.Treasure;

/**
 * Created by baoyongzhang on 2016/10/26.
 */
public class SettingsActivity extends AppCompatActivity {

    private TextView mSearchEngineText;
    private SwitchCompat mAutoCopySwitch;
    private SwitchCompat mListenClipboardSwitch;
    private Config mConfig;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_settings);

        findViewById(R.id.search_engine).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new AlertDialog.Builder(SettingsActivity.this).setItems(SearchEngine.getSupportSearchEngineList(), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mConfig.setSearchEngine(SearchEngine.getSupportSearchEngineList()[which]);
                        BigBang.registerAction(BigBang.ACTION_SEARCH, SearchEngine.getSearchAction(getApplicationContext()));
                        updateUI();
                    }
                }).show();
            }
        });

        mSearchEngineText = (TextView) findViewById(R.id.search_engine_text);

        mAutoCopySwitch = (SwitchCompat) findViewById(R.id.auto_copy_switch);
        mAutoCopySwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                mConfig.setAutoCopy(isChecked);
                BigBang.registerAction(BigBang.ACTION_BACK, mConfig.isAutoCopy() ? CopyAction.create() : null);
                updateUI();
            }
        });

        mListenClipboardSwitch = (SwitchCompat) findViewById(R.id.listen_clipboard_switch);
        mListenClipboardSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                mConfig.setListenClipboard(isChecked);
                if (mConfig.isListenClipboard()) {
                    ListenClipboardService.start(getApplicationContext());
                } else {
                    ListenClipboardService.stop(getApplicationContext());
                }
                updateUI();
            }
        });

        findViewById(R.id.bigbang_style).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(SettingsActivity.this, StyleActivity.class));
            }
        });
        mConfig = Treasure.get(this, Config.class);
        updateUI();

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    private void updateUI() {
        mSearchEngineText.setText(mConfig.getSearchEngine());
        mAutoCopySwitch.setChecked(mConfig.isAutoCopy());
        mListenClipboardSwitch.setChecked(mConfig.isListenClipboard());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/im/dacer/kata/SettingsActivity.kt
package im.dacer.kata

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import com.baoyz.treasure.Treasure
import im.dacer.kata.core.data.MultiprocessPref
import im.dacer.kata.core.util.LangUtils
import im.dacer.kata.core.util.WebParser
import im.dacer.kata.service.ListenClipboardService
import kotlinx.android.synthetic.main.activity_settings.*

class SettingsActivity : AppCompatActivity() {

    private val mConfig by lazy { Treasure.get(this, Config::class.java) }
    private val appPref by lazy { MultiprocessPref(this) }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)


        searchEngine.setOnClickListener {
            AlertDialog.Builder(this@SettingsActivity)
                    .setItems(SearchEngine.getSupportSearchEngineList()) { _, which ->
                appPref.searchEngine = SearchEngine.getSupportSearchEngineList()[which]
                updateUI()
            }.show()
        }

        enhancedModeSwitch.setOnCheckedChangeListener { _, isChecked ->
            appPref.enhancedMode = isChecked
            refreshService()
            updateUI()
        }

        translationTarget.setOnClickListener {
            AlertDialog.Builder(this@SettingsActivity)
                    .setItems(LangUtils.LANG_LIST) { _, which ->
                appPref.targetLang = LangUtils.LANG_KEY_LIST[which]
                updateUI()
            }.show()
        }

        webPageParser.setOnClickListener {
            AlertDialog.Builder(this@SettingsActivity)
                    .setItems(WebParser.getParserNameArray(this)) { _, which ->
                appPref.webParser = WebParser.Parser.values()[which]
                updateUI()
            }.show()
        }

        showFloatDialogSwit.setOnCheckedChangeListener { _, isChecked ->
            appPref.showFloatDialog = isChecked
            updateUI()
        }

        listenClipboardSwitch.setOnCheckedChangeListener { _, isChecked ->
            mConfig.isListenClipboard = isChecked
            refreshService()
            updateUI()
        }

        bigbangStyle.setOnClickListener { startActivity(Intent(this@SettingsActivity, StyleActivity::class.java)) }
        updateUI()

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }

    private fun refreshService() {
        if (mConfig.isListenClipboard) {
            ListenClipboardService.restart(applicationContext)
        } else {
            ListenClipboardService.stop(applicationContext)
        }
    }

    private fun updateUI() {
        searchEngineTv.text = appPref.searchEngine
        showFloatDialogSwit.isChecked = appPref.showFloatDialog
        listenClipboardSwitch.isChecked = mConfig.isListenClipboard
        webPageParserTv.setText(appPref.webParser.stringRes)
        enhancedModeSwitch.isChecked = appPref.enhancedMode
        translationTargetTv.text = LangUtils.getLangByKey(appPref.targetLang)

        enhancedModeSwitch.isEnabled = listenClipboardSwitch.isChecked
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }
}
